test/unit/test-sharedmem.cpp

```diff
@@ -0,0 +1,1032 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-18, Lawrence Livermore National Security, LLC.
+//
+// Produced at the Lawrence Livermore National Laboratory
+//
+// LLNL-CODE-689114
+//
+// All rights reserved.
+//
+// This file is part of RAJA.
+//
+// For details about use and distribution, please read RAJA/LICENSE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/RAJA.hpp"
+#include "RAJA_gtest.hpp"
+
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <cmath>
+#include <cassert>
+
+#include "camp/camp.hpp"
+#include "camp/concepts.hpp"
+
+#if defined(RAJA_ENABLE_CUDA)
+#include <cuda_runtime.h>
+#endif
+
+using namespace RAJA;
+using namespace RAJA::statement;
+
+//Define tile size ( TILE_DIM x TILE_DIM )
+//Matrix transpose and matrix multiplication
+//are carried out via tiling algorithms
+RAJA_INDEX_VALUE(TX, "TX");
+RAJA_INDEX_VALUE(TY, "TY");
+
+const int TILE_DIM = 16;
+
+template <typename NestedPolicy>
+class TypedLocalMem : public ::testing::Test
+{
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+};
+TYPED_TEST_CASE_P(TypedLocalMem);
+
+CUDA_TYPED_TEST_P(TypedLocalMem, Basic)
+{
+  using Pol = at_v<TypeParam, 0>;
+
+  const int DIM = 2;
+  const int N_rows = 144;
+  const int N_cols = 255;
+
+  const int inner_Dim0 = TILE_DIM;
+  const int inner_Dim1 = TILE_DIM;
+
+  const int outer_Dim0 = (N_cols-1)/TILE_DIM+1;
+  const int outer_Dim1 = (N_rows-1)/TILE_DIM+1;
+
+  double *A, *B;
+#if defined(RAJA_ENABLE_CUDA)
+  size_t Arr_sz = N_rows * N_cols;
+  cudaMallocManaged(&A,  sizeof(double) * Arr_sz);
+  cudaMallocManaged(&B, sizeof(double)  * Arr_sz);
+#else
+  A  = new double[N_rows * N_cols];
+  B  = new double[N_rows * N_cols];
+#endif
+
+  RAJA::View<double, RAJA::Layout<DIM>> Aview(A, N_rows, N_cols);
+  RAJA::View<double, RAJA::Layout<DIM>> Bview(B, N_rows, N_cols);
+
+  for (int row = 0; row < N_rows; ++row) {
+    for (int col= 0 ; col < N_cols; ++col) {
+      Aview(row, col) = col;
+    }
+  }
+
+  using SharedTile = TypedLocalArray<double, RAJA::SizeList<TILE_DIM,TILE_DIM>, TY, TX>;
```

Comment:
@artv3 here, you pass strongly-typed index args as template params to define the tile type, but then below you have 'int' args passed to the lambda and you manually cast them back to strongly-typed. Can't you have strongly-typed index args passed to the lambda?

Reply:
That's a good point, I just updated it.